torch/csrc/autograd/engine.cpp

```diff
@@ -308,37 +309,15 @@ auto Engine::thread_main(
   // Why the test on graph_task->outstanding_tasks_?  See
   // Note [Reentrant backwards]
   while (!reentrant_thread || graph_task->outstanding_tasks_ > 0) {
-    NodeTask task = local_ready_queue->pop();
+    auto task = local_ready_queue->pop();
     // This will only work if the worker is running a non backward task
     // TODO Needs to be fixed this to work in all cases
-    if (task.isShutdownTask_) {
+    if (task->isShutdownTask_) {
       C10_LOG_API_USAGE_ONCE("torch.autograd.thread_shutdown");
       break;
     }
-
-    // local_graph_task represents the graph_task we retrieve from the queue.
-    // The outer graph_task represents the overall graph_task we need to execute
-    // for reentrant execution.
-    std::shared_ptr<GraphTask> local_graph_task;
-    if (!(local_graph_task = task.base_.lock())) {
-      // If the task's graph_task expired, just ignore the task.
-      LOG(INFO) << "GraphTask for function " << task.fn_->name()
-                << " is no longer valid, skipping execution";
-      continue;
-    }
-
-    if (task.fn_ && !local_graph_task->has_error_.load()) {
-      AutoGradMode grad_mode(local_graph_task->grad_mode_);
-      try {
-        evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
-      } catch (std::exception& e) {
-        thread_on_exception(local_graph_task, task.fn_, e);
-      }
-    }
-
-    // Decrement the outstanding tasks.
-    --local_graph_task->outstanding_tasks_;
-
+    execute_node_task(task);
+    std::shared_ptr<GraphTask> local_graph_task = task->base_.lock();
```

Comment:
Can we update `execute_node_task` to return the local_graph_task to avoid doing this? Also, we need to check whether local_graph_task is nullptr here since `execute_node_task` might have just returned due to that error and we don't want to continue executing with a null local_graph_task